scio-smb/src/main/java/org/apache/beam/sdk/extensions/smb/AvroFileOperations.java

```diff
@@ -103,50 +82,40 @@ public void populateDisplayData(Builder builder) {
 
   @Override
   protected Reader<ValueT> createReader() {
-    return new AvroReader<>(recordClass, schemaSupplier);
+    return new AvroReader<>(datumFactory, schemaSupplier);
   }
 
   @SuppressWarnings("unchecked")
   @Override
   protected FileIO.Sink<ValueT> createSink() {
     final AvroIO.Sink<ValueT> sink =
-        recordClass == null
-            // https://github.com/spotify/scio/issues/2649
-            // force GenericDatumWriter instead of ReflectDatumWriter
-            ? (AvroIO.Sink<ValueT>)
-                AvroIO.<GenericRecord>sink(getSchema())
-                    .withDatumWriterFactory(AvroDatumFactory.generic())
-            : AvroIO.sink(recordClass)
-                .withDatumWriterFactory(
-                    (writer) -> {
-                      // same as SpecificRecordDatumFactory in scio-avro
-                      ReflectData data = new ReflectData(recordClass.getClassLoader());
-                      org.apache.beam.sdk.extensions.avro.schemas.utils.AvroUtils
-                          .addLogicalTypeConversions(data);
-                      return new ReflectDatumWriter<>(writer, data);
-                    });
+        ((AvroIO.Sink<ValueT>) AvroIO.sink(getSchema()))
+                .withDatumWriterFactory(datumFactory)
+                .withCodec(codec.getCodec());
 
     if (metadata != null) {
       return sink.withMetadata(metadata);
```

Comment:
I think there was a bug here: codec was not set if metadata was not null

Reply:
ack! good thing there wasn't much adoption of the the metadata feature here... 😬

Reply:
BTW, metadata is not exposed on the IO level. Is this expected ?

Reply:
it's not part of AvroSortedBucketIO? It should be 😬